NBitcoin/Protocol/NetworkAddress.cs

```diff
@@ -5,113 +5,383 @@
+		private byte[] addr = new byte[ADDR_IPV6_SIZE];
+
+		private ushort port;
+
+		public ushort Port => port;
+
+		static uint TIME_INIT = 100000000;
+
+		// disk and network only
+		internal uint nTime = TIME_INIT;
+
+		private ulong services = (ulong)NodeServices.Nothing;
+
 		public NetworkAddress()
 		{
+		}
 
+		public NetworkAddress(IPAddress ip, int port)
+			: this(ip)
+		{
+			this.port = (ushort)port;
 		}
-		public NetworkAddress(IPEndPoint endpoint)
+
+		public NetworkAddress(IPAddress ip)
 		{
-			Endpoint = endpoint;
+			SetIp(ip);
 		}
-		public NetworkAddress(IPAddress address, int port)
+
+		public NetworkAddress(EndPoint endPoint)
 		{
-			Endpoint = new IPEndPoint(address, port);
+			Endpoint = endPoint;;
 		}
 
-		internal uint ntime;
-		ulong service = 1;
-		byte[] ip = new byte[16];
-		ushort port;
+		public bool IsIPv4 => (Network)network == Network.IPv4;
+		public bool IsIPv6 => (Network)network == Network.IPv6;
+		public bool IsTor => (Network)network == Network.Onion;
+		public bool IsI2P => (Network)network == Network.I2P;
+		public bool IsCjdns => (Network)network == Network.Cjdns;
 
-		public ulong Service
+		public EndPoint Endpoint
 		{
 			get
 			{
-				return service;
+				if (IsAddrV1Compatible)
+				{
+					return new IPEndPoint(new IPAddress(SerializeV1Array()), port);
+				}
+				else
+				{
+					return new DnsEndPoint(ToAddressString(), port);
+				}
 			}
 			set
 			{
-				service = value;
+				if (value is IPEndPoint ipEndPoint)
+				{
+					SetIp(ipEndPoint.Address);
+					port = (ushort)ipEndPoint.Port;
+				}
+				else if (value is DnsEndPoint dnsEndPoint)
+				{
+					SetSpecial(dnsEndPoint.Host);
+					port = (ushort)dnsEndPoint.Port;
+				}
+				else
+				{
+					throw new ArgumentException($"Not supported {nameof(EndPoint)} type {value.GetType().FullName}.");
+				}
 			}
+		} 
+
+		protected void SetIp(IPAddress ip)
+		{
+			SetLegacyIpv6(ip.MapToIPv6().GetAddressBytes());
 		}
 
-		public TimeSpan Ago
+		public void SetLegacyIpv6(byte[] ipv6)
 		{
-			get
+			if (ipv6.Length != ADDR_IPV6_SIZE)
 			{
-				return DateTimeOffset.UtcNow - Time;
+				throw new ArgumentException("The received value is not a valid IPv6 byte array.", nameof(ipv6));
 			}
-			set
+
+			var skip = 0;
+			if (HasPrefix(ipv6, IPV4_IN_IPV6_PREFIX))
 			{
-				Time = DateTimeOffset.UtcNow - value;
+				network = (byte)Network.IPv4;
+				skip = IPV4_IN_IPV6_PREFIX.Length;
 			}
-		}
+			else if(HasPrefix(ipv6, TORV2_IN_IPV6_PREFIX))
+			{
+				// TORv2-in-IPv6
+				network = (byte)Network.Onion;
+				skip = TORV2_IN_IPV6_PREFIX.Length;
+			} 
+			else
+			{
+				// IPv6
+				network = (byte)Network.IPv6;
+			}	
 
-		public void Adjust()
-		{
-			var nNow = Utils.DateTimeToUnixTime(DateTimeOffset.UtcNow);
-			if (ntime <= 100000000 || ntime > nNow + 10 * 60)
-				ntime = nNow - 5 * 24 * 60 * 60;
+			addr = ipv6.Skip(skip).ToArray();
 		}
 
-		public IPEndPoint Endpoint
+		public bool SetSpecial(string address)

```

Comment:
I have to think about this.

Reply:
Ok, I've just push a PR to the bitcoin repo https://github.com/bitcoin/bitcoin/pull/20662. Let see what the core team thinks about it. In case the PR is ACKed then I will apply the same changes here; otherwise I will rename it as `SetOnionAddress` as you suggest because that will be a much better name.

Reply:
I will support I2P here too because even if my PR is not finally merged there is another one (more ambitious) https://github.com/bitcoin/bitcoin/pull/20254 that has changes to be merged and contains the exact same changes.

Reply:
Done.